crates/core/src/bc_protocol/users.rs

```diff
@@ -0,0 +1,229 @@
+                .await?;
+            let get = Bc {
+                meta: BcMeta {
+                    msg_id: MSG_ID_GET_ABILITY_SUPPORT,
+                    channel_id: self.channel_id,
+                    msg_num,
+                    response_code: 0,
+                    stream_type: 0,
+                    class: 0x6414,
+                },
+                body: BcBody::ModernMsg(ModernMsg {
+                    extension: Some(Extension {
+                        user_name: Some("admin".to_owned()),
+                        ..Default::default()
+                    }),
+                    payload: None,
+                }),
+            };
+
+            sub_get.send(get).await?;
+            let msg = sub_get.recv().await?;
+            if msg.meta.response_code == 400 {
+                // Retryable
+                if reties < 5 {
+                    reties += 1;
+                    continue;
+                } else {
+                    return Err(Error::CameraServiceUnavailable {
+                        id: msg.meta.msg_id,
+                        code: msg.meta.response_code,
+                    });
+                }
+            } else if msg.meta.response_code != 200 {
+                return Err(Error::CameraServiceUnavailable {
+                    id: msg.meta.msg_id,
+                    code: msg.meta.response_code,
+                });
+            } else {
+                // Valid message with response_code == 200
+                if let BcBody::ModernMsg(ModernMsg {
+                    payload:
+                        Some(BcPayloads::BcXml(BcXml {
+                            user_list: Some(user_list),
+                            ..
+                        })),
+                    ..
+                }) = msg.body
+                {
+                    return Ok(user_list);
+                } else {
+                    return Err(Error::UnintelligibleReply {
+                        reply: std::sync::Arc::new(Box::new(msg)),
+                        why: "Expected ModernMsg payload with a user_list but it was not recieved",
+                    });
+                }
+            }
+        }
+    }
+
+    /// Add a new user
+    ///
+    /// This function does not check if the user exist and the API will likely throw an error if
+    /// that is the case
+    pub async fn add_user(
+        &self,
+        user_name: String,
+        password: String,
+        user_level: u8,
+    ) -> Result<()> {
+        let users = self.get_users().await;
+
+        let mut users = match users {
+            Ok(UserList {
+                user_list: Some(users),
+                ..
+            }) => users,
+            _ => Vec::new(),
+        };
+
+        users.push(User {
+            user_set_state: "add".to_owned(),
+            user_name,
+            password: Some(password),
+            user_level,
+            user_id: None,
+            login_state: None,
+        });
```

Comment:
```suggestion
        if users.iter().find(|user| user.user_name == user_name).is_some() {
            return Err(Error::Other("User already exists"));
        } else {        
          users.push(User {
              user_set_state: "add".to_owned(),
              user_name,
              password: Some(password),
              user_level,
              user_id: None,
              login_state: None,
          });
        }
```

Reply:
cargo clippy suggested `.any(...)` over `.find(...).is_some()`

Reply:
yes `.any(...)` (I was trying to remember that one, thank you clippy)